pkg/api/validate/dynamic/serviceprincipal_test.go

```diff
@@ -0,0 +1,193 @@
+package dynamic
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"crypto/hmac"
+	"crypto/sha512"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"testing"
+
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/golang/mock/gomock"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/util/azureclient"
+	mock_aad "github.com/Azure/ARO-RP/pkg/util/mocks/aad"
+)
+
+type tokenRequirements struct {
+	clientID      string
+	clientSecret  string
+	tenantID      string
+	aadEndpoint   string
+	graphEndpoint string
+	resource      string
+	claims        string
+	signMethod    string
+}
+
+func TestValidateServicePrincipal(t *testing.T) {
+	ctx := context.Background()
+	log := logrus.NewEntry(logrus.StandardLogger())
+
+	for _, tt := range []struct {
+		tokenReqs *tokenRequirements
+		name      string
+		wantErr   string
+		aadMock   func(*mock_aad.MockTokenClient, *tokenRequirements)
+	}{
+		{
+			name:      "pass: Successful Validation",
+			tokenReqs: newTokenRequirements(),
+			aadMock: func(aad *mock_aad.MockTokenClient, tr *tokenRequirements) {
+				token, err := createToken(tr)
+				if err != nil {
+					t.Errorf("failed to create testing service principal token: %v\n", err)
+				}
+				aad.EXPECT().GetToken(ctx, log, tr.clientID, tr.clientSecret, tr.tenantID, tr.aadEndpoint, tr.graphEndpoint).MaxTimes(1).Return(token, nil)
+			},
```

Comment:
There is lots of code duplication here, I think it would be better to have that code in the test section and only feed it the arguments

Reply:
Thanks Jeremy. I originally did that to change the error returned. Now it seems obvious to create an another error field in the test struct. I moved the mock code in my last commit.